artifact/tar/tar.hpp

```diff
@@ -0,0 +1,104 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_TAR_HPP
+#define MENDER_COMMON_TAR_HPP
+
+#include <string>
+#include <vector>
+
+#include <config.h>
+
+#include <common/io.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+
+#ifdef MENDER_TAR_LIBARCHIVE
+#include <libarchive/wrapper.hpp>
+#endif
+
+namespace mender {
+namespace tar {
+
+using namespace std;
+
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace io = mender::common::io;
+
+
+enum ErrorCode {
+	NoError = 0,
+	TarReaderError,
+	TarShortReadError,
+	TarEntryError,
+	TarEOFError,
+};
+
+class ErrorCategoryClass : public std::error_category {
+public:
+	const char *name() const noexcept override;
+	string message(int code) const override;
+};
+
+extern const ErrorCategoryClass ErrorCategory;
+
+error::Error MakeError(ErrorCode code, const string &msg);
+
+class Entry : public io::Reader {
+private:
+	std::string name_;
+	uint64_t total_size_;
+
+	Reader &reader_;
+
+	// Reader data
+	size_t nr_bytes_read_ {0};
+
+public:
+	Entry(std::string name, uint64_t archive_size, Reader &reader) :
+		name_ {name},
+		total_size_ {archive_size},
+		reader_ {reader} {
+	}
+
+	std::string Name() {
+		return name_;
+	}
+
+	expected::ExpectedSize Read(
+		vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) override;
+};
+
+using ExpectedEntry = expected::Expected<Entry, error::Error>;
+
+class Reader : io::Reader {
+private:
+#ifdef MENDER_TAR_LIBARCHIVE
+	mender::libarchive::wrapper::Handle archive_handle;
+#endif
+
+	expected::ExpectedSize Read(
+		vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) override;
+
+public:
+	Reader(mender::common::io::Reader &reader);
+
+	ExpectedEntry Next();
```

Comment:
Could we move this `Next()` to the `Handle` class and get rid of this `Reader` class?

Reply:
I don't think we can get completely rid of it, something needs to own the `unique_ptr`, or rather, the `handle` in this instance.

Reply:
I don't follow, sorry. This holds an instance of `Handle`. It actually wraps the `Handle` and provides one extra function, `Next()` on top of it. So you can either move this `Next()` function to the `Handle` class and get rid of this `Reader` class or you can base this class on top of the `Handle` class (inherit it) and then only implement the extra `Next()` function here.

Reply:
Solved f2f. Leaving it be for now. It is OK as is.